### PR TITLE
Add enough arbitraries to test SemigroupK[EntityDecoder[F, ?]]

### PR DESCRIPTION
**Pull request by rossabaker**
Continuation of #1488.  Adds the test for what I prematurely merged.

/cc @jadlr

### CODE REVIEW

File: core/src/main/scala/org/http4s/MediaType.scala [hunk 1]
```diff
@@ -145,10 +142,15 @@ object MediaRange {
       override def render(writer: Writer, range: MediaRange): writer.type =
         range.render(writer)
 
-      override def eqv(x: MediaRange, y: MediaRange): Boolean =
-        x.equals(y)
+      override def compare(x: MediaRange, y: MediaRange): Int = {
+        def orderedSubtype(a: MediaRange) = a match {
+          case mt: MediaType => mt.subType
+          case _ => ""
+        }
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/MediaType.scala]: The awkwardness of this flows from the awkwardness of the types.  We should do better here.  (Sealed concrete types? :confused:)

@cquiroz mentioned he's been thinking about this.

File: core/src/main/scala/org/http4s/MessageFailure.scala [hunk 2]
```diff
@@ -70,6 +70,10 @@ object ParseResult {
   * the entity headers or with the entity itself.   */
 trait DecodeFailure extends MessageFailure
 
+object DecodeFailure {
+  implicit val http4sEqForDecodeFailure: Eq[DecodeFailure] = Eq.fromUniversalEquals
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/MessageFailure.scala]: I don't feel great about this instance.

File: core/src/main/scala/org/http4s/MessageFailure.scala [hunk 3]
```diff
@@ -96,7 +100,7 @@ final case class InvalidMessageBodyFailure(details: String, cause: Option[Throwa
       .withBody(s"The request body was invalid.")(F, EntityEncoder.stringEncoder[F])
 }
 
-/** Indicates that a [[Message]] came with no supported [[MediaType]]. */
+/** Indicates that Message]] came with no supported [[MediaType]]. */
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/MessageFailure.scala]: Oops.  Will fix.

File: testing/src/main/scala/org/http4s/testing/ArbitraryInstances.scala [hunk 4]
```diff
@@ -267,22 +267,36 @@ trait ArbitraryInstances {
...
-    }
+    Arbitrary(http4sGenMediaRange)
 
   implicit val http4sCogenMediaRange: Cogen[MediaRange] =
-    Cogen[(String, Map[String, String])].contramap(m => (m.mainType, m.extensions))
+    Cogen[(String, String, Map[String, String])].contramap { m =>
+      val effectiveSubtype = m match {
+        case mt: MediaType => mt.subType
+        case _ => "*"
+      }
```
Comment (rossabaker) [inline testing/src/main/scala/org/http4s/testing/ArbitraryInstances.scala]: More garbage from the bad structure of `MediaType` and `MediaRange`.

File: testing/src/main/scala/org/http4s/testing/ArbitraryInstances.scala [hunk 5]
```diff
@@ -350,7 +364,7 @@ trait ArbitraryInstances {
   implicit val arbitraryContentLength: Arbitrary[`Content-Length`] =
     Arbitrary {
       for {
-        long <- arbitrary[Long] if long > 0L
+        long <- sized(n => choose(0L, n.toLong))
```
Comment (rossabaker) [inline testing/src/main/scala/org/http4s/testing/ArbitraryInstances.scala]: The filter could cause this generator to fail on lists of headers.  This fix should be backported to 0.18.

File: testing/src/main/scala/org/http4s/testing/ArbitraryInstances.scala [hunk 6]
```diff
@@ -715,9 +729,100 @@ trait ArbitraryInstances {
...
+  ) =
+    Arbitrary(for {
+      f <- arbitrary[(Message[F], Boolean) => DecodeResult[F, A]]
+      mrs <- arbitrary[Set[MediaRange]]
+    } yield new EntityDecoder[F, A] {
+      def decode(msg: Message[F], strict: Boolean): DecodeResult[F, A] = f(msg, strict)
+      def consumes = mrs
+    })
+
+  implicit def cogenMessage[F[_]](implicit F: Effect[F], ec: TestContext): Cogen[Message[F]] =
```
Comment (rossabaker) [inline testing/src/main/scala/org/http4s/testing/ArbitraryInstances.scala]: Incomplete.  It doesn't account for the differences between requests and responses.

File: testing/src/main/scala/org/http4s/testing/ArbitraryInstances.scala [hunk 7]
```diff
@@ -715,9 +729,100 @@ trait ArbitraryInstances {
...
+    Cogen[(Headers, EntityBody[F])].contramap(m => (m.headers, m.body))
+
+  implicit def cogenHeaders: Cogen[Headers] =
+    Cogen[List[Header]].contramap(_.toList)
+
+  implicit def cogenHeader: Cogen[Header] =
+    Cogen[(CaseInsensitiveString, String)].contramap(h => (h.name, h.value))
+
+  implicit def arbitraryDecodeFailure: Arbitrary[DecodeFailure] =
```
Comment (rossabaker) [inline testing/src/main/scala/org/http4s/testing/ArbitraryInstances.scala]: Incomplete.  `DecodeFailure` is an open hierarchy.  Which I dearly wish it weren't.  If we do complete this, then we have a problem in the Cogen instance.

File: testing/src/main/scala/org/http4s/testing/ArbitraryInstances.scala [hunk 8]
```diff
@@ -715,9 +729,100 @@ trait ArbitraryInstances {
...
+        case MediaTypeMismatch(mt, e) => Cogen[(MediaType, Set[MediaRange])].perturb(seed, (mt, e))
+        // TODO What if it's not one of these?
+      }
+    }
+
+  private[http4s] implicit def arbitraryMessage[F[_]]: Arbitrary[Message[F]] =
+    // TODO this is bad because the underlying generators are bad
+    Arbitrary(Gen.oneOf(arbitrary[Request[F]], arbitrary[Response[F]]))
+
+  private[http4s] implicit def arbitraryRequest[F[_]]: Arbitrary[Request[F]] = Arbitrary {
```
Comment (rossabaker) [inline testing/src/main/scala/org/http4s/testing/ArbitraryInstances.scala]: This, and response, are something I've wanted for a long time, but they need to get a lot more sophisticated to make sure they're following the HTTP rules.

File: tests/src/test/scala/org/http4s/EntityDecoderSpec.scala [hunk 9]
```diff
@@ -15,6 +21,10 @@ import scala.concurrent.ExecutionContext
 
 class EntityDecoderSpec extends Http4sSpec with PendingUntilFixed {
   implicit val executionContext: ExecutionContext = trampoline
+  implicit val testContext: TestContext = TestContext()
+
+  implicit def entityDecoderEq[A: Eq]: Eq[EntityDecoder[IO, A]] =
```
Comment (rossabaker) [inline tests/src/test/scala/org/http4s/EntityDecoderSpec.scala]: Maybe we should have an `eq` package in our testing package for things that are "eq by lack of a counterexample," like cats does.

File: tests/src/test/scala/org/http4s/EntityDecoderSpec.scala [hunk 10]
```diff
@@ -439,4 +449,5 @@ class EntityDecoderSpec extends Http4sSpec with PendingUntilFixed {
     EntityEncoder.simple[IO, ErrorJson](`Content-Type`(MediaType.`application/json`))(json =>
       Chunk.bytes(json.value.getBytes()))
 
+  checkAll("SemigroupK[EntityDecoder[IO, ?]]", SemigroupKTests[EntityDecoder[IO, ?]].semigroupK[String])
```
Comment (rossabaker) [inline tests/src/test/scala/org/http4s/EntityDecoderSpec.scala]: The line that motivated this entire PR.  Sad: it takes 81 seconds on my machine.

File: tests/src/test/scala/org/http4s/MediaRangeSpec.scala [hunk 11]
```diff
@@ -41,6 +42,6 @@ class MediaRangeSpec extends Http4sSpec {
     }
   }
 
-  checkAll("Eq[MediaRange]", EqTests[MediaRange].eqv)
+  checkAll("Order[MediaRange]", OrderTests[MediaRange].order)
```
Comment (rossabaker) [inline tests/src/test/scala/org/http4s/MediaRangeSpec.scala]: Was necessary so we could get a Cogen for a set of them.  HTTP doesn't have anything to say about it, and maybe it should be demoted to a test package.

File: core/src/main/scala/org/http4s/MessageFailure.scala (hunk 2 above)
Comment (jmcardon) [reply core/src/main/scala/org/http4s/MessageFailure.scala]: Me neither, but then again I'm not exactly sure whether we care about matching too much whether two decodefailures match, vs simply knowing something failed in a specific way.

**Pull request merged**